source/fuzz/transformation_push_id_through_variable.cpp

```diff
@@ -0,0 +1,137 @@
+  if (!fuzzerutil::IsFreshId(ir_context, message_.value_synonym_id())) {
+    return false;
+  }
+
+  // The variable instruction must exist and have a type.
+  auto variable_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.variable_id());
+  if (!variable_instruction || !variable_instruction->type_id()) {
+    return false;
+  }
+
+  // The variable type instruction must be defined and be a OpTypePointer
+  // instruction.
+  auto variable_type_instruction =
+      ir_context->get_def_use_mgr()->GetDef(variable_instruction->type_id());
+  if (!variable_type_instruction ||
+      variable_type_instruction->opcode() != SpvOpTypePointer) {
+    return false;
+  }
+
+  // The variable must not be read-only.
+  if (variable_instruction->IsReadOnlyPointer()) {
+    return false;
+  }
+
+  // The instruction to insert before must be defined.
+  auto instruction_to_insert_before =
+      FindInstruction(message_.instruction_descriptor(), ir_context);
+  if (!instruction_to_insert_before) {
+    return false;
+  }
+
+  // The instruction to insert before must belongs to a reachable block.
+  if (!fuzzerutil::BlockIsReachableInItsFunction(
+          ir_context,
+          ir_context->get_instr_block(instruction_to_insert_before))) {
+    return false;
+  }
+
+  // It must be valid to insert the OpStore and OpLoad instruction before it.
+  if (!fuzzerutil::CanInsertOpcodeBeforeInstruction(
+          SpvOpStore, instruction_to_insert_before) ||
+      !fuzzerutil::CanInsertOpcodeBeforeInstruction(
+          SpvOpLoad, instruction_to_insert_before)) {
+    return false;
+  }
+
+  // The value instruction must be defined and have a type.
+  auto value_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.value_id());
+  if (!value_instruction || !value_instruction->type_id()) {
+    return false;
+  }
+
+  // The variable pointee type must be equal to the value type.
+  if (variable_type_instruction->GetSingleWordInOperand(1) !=
+      value_instruction->type_id()) {
+    return false;
+  }
+
+  // |message_.variable_id| must be available at the insertion point.
+  if (!fuzzerutil::IdIsAvailableBeforeInstruction(
+          ir_context, instruction_to_insert_before, message_.variable_id())) {
+    return false;
+  }
+
+  // |message_.value_id| must be available at the insertion point.
+  return fuzzerutil::IdIsAvailableBeforeInstruction(
+      ir_context, instruction_to_insert_before, message_.value_id());
+}
+
+void TransformationPushIdThroughVariable::Apply(
+    opt::IRContext* ir_context,
+    TransformationContext* transformation_context) const {
+  TransformationStore(message_.variable_id(), message_.value_id(),
+                      message_.instruction_descriptor())
+      .Apply(ir_context, transformation_context);
+
+  TransformationLoad(message_.value_synonym_id(), message_.variable_id(),
+                     message_.instruction_descriptor())
+      .Apply(ir_context, transformation_context);
```

Comment:
@paulthomson What do you think of this?  We haven't done this before - used Apply methods from other transformations in a larger transformation.

Reply:
This seems OK. Perhaps it is slightly confusing, but a comment will make it clear that the transformations are not being added to the list of transformations. Apply might update the fact manager, but that is OK too I think.

Reply:
But actually, in this particular case, I think it is clearer to just add the code that adds the load and store instructions because there is not much code.

Reply:
Done.